## Python Exercises
##### Ben Steves
##### DATA 303
##### 3-28-22

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error
import string as str
import sklearn

In [2]:
words = pd.read_csv("https://raw.githubusercontent.com/seanpatlan/wordle-words/main/word-bank.csv", header=None)

In [3]:
words.head()

0
0  aback
1  abase
2  abate
3  abbey
4  abbot

In [4]:
words.rename(columns = {0:'word'}, inplace = True)

In [5]:
words.head(10)

word
0  aback
1  abase
2  abate
3  abbey
4  abbot
5  abhor
6  abide
7  abled
8  abode
9  abort

### 1 A

In [6]:
len(words)

2315

There are 2315 words in this file

### 1B

In [7]:
words['startsQ'] = words['word'].str.startswith("q")
len(words[words['startsQ']==True])

23

There are 23 words that start with Q

### 1C

In [8]:
#function returns true if repeated letters
#false otherwise
def repeats(string):
  duplicates = []
  for i in range(len(string)):
    if string[i] in duplicates:
      return True
      break
    duplicates.append(string[i])
  return False

In [9]:
words['duplicate'] = words['word'].apply(repeats)
len(words[words['duplicate']==True])

749

There are 749 words with one repeated letter

### 1D

In [10]:
#using this function for next two questions

''' checkWordsForLetters returns a matrix. Adds the value 1 to matrix indexes where a letter 
from input  "letters" is present in a word from input "words", and a 0 is added otherwise'''
def checkWordsForLetters(words, letters):

  #create empty matrix w/ rows as the words and columns as letters
  matrix = np.empty((len(words), len(letters)))

  for i in range(len(words)): #for each word
    for j in range(len(letters)): #for each letter
      if letters[j] in words[i]: #check if letter at index j is in the current word
        matrix[i, j] = 1 #for this word, add a 1 each time a matching letter is found
      else:
        matrix[i, j] = 0 #add 0 if no match

  return matrix


In [11]:
word_list = words['word'].to_list() #get word values
letters = 'abcdefghijklmnopqrstuvwxyz' #storing letters in a string
matrix1 = checkWordsForLetters(word_list, letters)

In [12]:
#create df and use letters as columns and add to matrix
letter_sums = pd.DataFrame(matrix1, columns = list(letters)) 

#merely to check correct word maps to correct matrix values
letter_sums['word'] = word_list 

#sum by letter and sort
letter_sums = letter_sums.drop('word', axis=1).sum(axis=0)
print(letter_sums.sort_values(ascending=False), '\n',(letter_sums/len(word_list)).sort_values(ascending=False))

e    1056.0
a     909.0
r     837.0
o     673.0
t     667.0
l     648.0
i     647.0
s     618.0
n     550.0
u     457.0
c     448.0
y     417.0
h     379.0
d     370.0
p     346.0
g     300.0
m     298.0
b     267.0
f     207.0
k     202.0
w     194.0
v     149.0
x      37.0
z      35.0
q      29.0
j      27.0
dtype: float64 
 e    0.456156
a    0.392657
r    0.361555
o    0.290713
t    0.288121
l    0.279914
i    0.279482
s    0.266955
n    0.237581
u    0.197408
c    0.193521
y    0.180130
h    0.163715
d    0.159827
p    0.149460
g    0.129590
m    0.128726
b    0.115335
f    0.089417
k    0.087257
w    0.083801
v    0.064363
x    0.015983
z    0.015119
q    0.012527
j    0.011663
dtype: float64


### 1E

I'm going to return words without traditional vowels. This is in response to me not getting "nymph" from a few days ago.

In [13]:
#similar idea to last one, except count across the columns this time
vowels = 'aeiou'
matrix2 = checkWordsForLetters(word_list, vowels)

no_vowel_words = pd.DataFrame(matrix2, columns = list(vowels))
no_vowel_words['word'] = word_list

no_vowel_words['unique_vowels'] = no_vowel_words.drop('word', axis=1).sum(axis=1)
no_vowel_words[no_vowel_words['unique_vowels'] == 0]

a    e    i    o    u   word  unique_vowels
499   0.0  0.0  0.0  0.0  0.0  crypt            0.0
609   0.0  0.0  0.0  0.0  0.0  dryly            0.0
875   0.0  0.0  0.0  0.0  0.0  glyph            0.0
945   0.0  0.0  0.0  0.0  0.0  gypsy            0.0
1174  0.0  0.0  0.0  0.0  0.0  lymph            0.0
1175  0.0  0.0  0.0  0.0  0.0  lynch            0.0
1283  0.0  0.0  0.0  0.0  0.0  myrrh            0.0
1320  0.0  0.0  0.0  0.0  0.0  nymph            0.0
1503  0.0  0.0  0.0  0.0  0.0  pygmy            0.0
1749  0.0  0.0  0.0  0.0  0.0  shyly            0.0
1799  0.0  0.0  0.0  0.0  0.0  slyly            0.0
2128  0.0  0.0  0.0  0.0  0.0  tryst            0.0
2305  0.0  0.0  0.0  0.0  0.0  wryly            0.0

### 2A

In [14]:
#note: I have loaded the data locally. Was having issues with the link
cars = pd.read_csv('data.csv')

In [15]:
cars.shape

(11914, 16)

There are 11914 rows and 16 columns

### 2B

Object types posted below. There's mix of object (categorical), integer, and float data types.

In [16]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

### 2C

In [17]:
np.mean(cars['MSRP'])

40594.737032063116

The mean of the variable MSRP is about 40594.74

### 2D

In [18]:
#cars

In [19]:
makes = cars.groupby('Make')['MSRP']
makes.mean().sort_values(ascending = False)

Make
Bugatti          1.757224e+06
Maybach          5.462219e+05
Rolls-Royce      3.511306e+05
Lamborghini      3.315673e+05
Bentley          2.471693e+05
McLaren          2.398050e+05
Ferrari          2.382188e+05
Spyker           2.133233e+05
Aston Martin     1.979104e+05
Maserati         1.142077e+05
Porsche          1.016224e+05
Tesla            8.525556e+04
Mercedes-Benz    7.147623e+04
Lotus            6.918828e+04
Land Rover       6.782322e+04
Alfa Romeo       6.160000e+04
BMW              6.154676e+04
Cadillac         5.623132e+04
Audi             5.345211e+04
Lexus            4.754907e+04
Genesis          4.661667e+04
Lincoln          4.283983e+04
Infiniti         4.239421e+04
HUMMER           3.646441e+04
Acura            3.488759e+04
GMC              3.049330e+04
Toyota           2.903002e+04
Nissan           2.858343e+04
Volvo            2.854116e+04
Chevrolet        2.835039e+04
Buick            2.820661e+04
Volkswagen       2.810238e+04
Saab             2.741350e+04
Ford 

### 2E

In [20]:
cars['Number of Doors'].fillna(value = 0,
          inplace = True)

cars['VehicleSize_tfm'] = LabelEncoder().fit_transform(cars['Vehicle Size'])

X_a = cars[['VehicleSize_tfm', 'Year']]
X_b = cars[['Number of Doors', 'highway MPG']]
y = cars['MSRP']

In [21]:
#same as 2D but listed w/ encoded values
makes_encoded = cars.groupby(['Vehicle Size', 'VehicleSize_tfm'])['MSRP']
makes_encoded.mean().sort_values(ascending = False)

Vehicle Size  VehicleSize_tfm
Large         1                  53890.500540
Midsize       2                  39035.919049
Compact       0                  34275.336482
Name: MSRP, dtype: float64

In [22]:
#make and year
X_train_a, X_test_a, y_train_a, y_test_a = train_test_split(X_a, y, test_size=0.20, random_state=34)

#num doors and mpg
X_train_b, X_test_b, y_train_b, y_test_b = train_test_split(X_b, y, test_size=0.20, random_state=34)

model1 = LinearRegression().fit(X_train_a, y_train_a)
model2 = LinearRegression().fit(X_train_b, y_train_b)

In [23]:
print(model1.coef_, model1.intercept_)
print(model2.coef_, model2.intercept_)

[ 518.72907609 1787.2434366 ] -3553297.4850351126
[-7941.41306811  -926.434174  ] 92436.17898572952


Model 1: MSRP = 518.73 * Size + 1787.2 * Year - 3553297.5

Model 2: MSRP = -7941.4 * NumDoors - 926.4 * highwayMPG + 92436.2

### 2F

In [24]:
m1_preds_train = model1.predict(X_train_a.iloc[[0]]) #Size: Midsize, Year: 1998
m1_preds_test = model1.predict(X_test_a.iloc[[0]]) #Size: Compact, Year: 2016
m2_preds_train = model2.predict(X_train_b.iloc[[0]]) #NumDoors: 4, MPG: 23
m2_preds_test = model2.predict(X_test_b.iloc[[0]]) #NumDoors: 2, MPG: 21

print('Model 1 w/ train: %f \nModel 1 w/ test: %f \nModel 2 w/ train: %f \nModel 2 w/ test: %f' 
      % (m1_preds_train, m1_preds_test, m2_preds_train, m2_preds_test))

#X_train_a.iloc[[0]]

Model 1 w/ train: 18652.359444 
Model 1 w/ test: 49785.283150 
Model 2 w/ train: 39362.540711 
Model 2 w/ test: 57098.235195


### 2G

In [25]:
#make predictions for all rows
y_pred_m1 = model1.predict(X_train_a)
y_pred_m2 = model2.predict(X_train_b)

print('MSE m1: %f, MSE m2: %f'
%(mean_squared_error(y_train_a, y_pred_m1), mean_squared_error(y_train_b, y_pred_m2)))

MSE m1: 3262659199.973437, MSE m2: 3314060925.421279


The MSE's are super high interestingly, but overall the first model fits the data better. Better is a choice term, though, because the MSE is huge. I used Mean Squared Error (MSE) which squares the residuals and takes the average.

### 2H

In [31]:
#make predictions for all rows
y_pred_m1 = model1.predict(X_test_a)
y_pred_m2 = model2.predict(X_test_b)

print('MSE m1: %f, MSE m2: %f'
%(mean_squared_error(y_test_a, y_pred_m1), mean_squared_error(y_test_b, y_pred_m2)))



MSE m1: 4075489791.712329, MSE m2: 4117719507.128470


8509    34815
Name: MSRP, dtype: int64

Model 1 works better on the test data as well. Expectedly, both of the models fit better on the training data.